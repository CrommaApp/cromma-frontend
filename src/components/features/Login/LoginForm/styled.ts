import styled from 'styled-components';

export const LoginFormContainer = styled.form`
	width: 100%;
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	padding: 10px 60px;
	overflow-y: scroll;
	margin-top: 4%;

	& > ul {
		width: 100%;
		display: flex;
		flex-direction: column;
		align-items: center;
		justify-content: center;
	}

	& > button {
		width: 100%;
		border: 1px solid #dddddd;
		background-color: white;
		padding: 3% 4%;
		border-radius: 4px;
		color: #888888;
		margin-top: 12%;
	}
`;

export const LoginFormInput = styled.li`
	width: 100%;
	margin: 10px 0;

	& > input {
		width: 100%;
		border: 1px solid #dddddd;
		border-radius: 4px;
		padding: 10px;
	}

	& > p {
		font-size: 0.7rem;
		color: #aaaaaa;
		margin: 0;
		padding: 4px;
	}
`;
