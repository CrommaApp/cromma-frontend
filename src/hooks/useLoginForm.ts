import React, { useState } from 'react';
import AuthService from '@services/auth/auth-service';
import { errorStatusState, successStatusState } from '@stores/status';
import { userState } from '@stores/user';
import { useSetRecoilState } from 'recoil';

const authService = new AuthService();

type LoginFormInputs = {
	id: string;
	password: string;
	closeLoginModal: () => void;
};

type ReturnTypes = [boolean, boolean, boolean, (e: React.FormEvent<HTMLFormElement>) => Promise<void>];

const useLoginForm = ({ id, password, closeLoginModal }: LoginFormInputs): ReturnTypes => {
	const [isIdValid, setIsIdValid] = useState(false);
	const [isPasswordValid, setIsPasswordValid] = useState(false);

	const [isFormSubmitted, setIsFormSubmitted] = useState(false);

	const setUser = useSetRecoilState(userState);
	const setErrorStatus = useSetRecoilState(errorStatusState);
	const setSuccessStatus = useSetRecoilState(successStatusState);

	const login = async () => {
		try {
			const result = await authService.login({
				userId: id,
				password,
			});

			if (result.statusCode === 200) {
				setUser((prev) => {
					return {
						...prev,
						isLogin: true,
						id: id,
					};
				});

				setSuccessStatus({ successMessage: result.message });

				closeLoginModal();
			}
		} catch (error: any) {
			setErrorStatus({ errorMessage: error.response?.data?.message || '잠시 후 다시 시도해주세요.' });
		}
	};

	const checkInputValidaton = () => {
		if (!id.trim() || !password.trim()) return false;

		const regExp = /^[A-Za-z0-9+]{4,10}$/;

		if (regExp.test(id)) {
			setIsIdValid(true);
		}

		if (regExp.test(password)) {
			setIsPasswordValid(true);
		}

		if (!regExp.test(id) || !regExp.test(password)) return false;

		return true;
	};

	const submitLoginForm = async (e: React.FormEvent<HTMLFormElement>) => {
		e.preventDefault();

		const result = checkInputValidaton();

		setIsFormSubmitted(true);

		if (result) {
			try {
				const result = await authService.signup({
					userId: id,
					password,
				});

				if (result.statusCode === 201) {
					await login();
				}
			} catch (error: any) {
				if (error.response.data.message === '이미 사용 중인 아이디입니다.') {
					await login();
				}
			}
		}
	};

	return [isIdValid, isPasswordValid, isFormSubmitted, submitLoginForm];
};

export default useLoginForm;
