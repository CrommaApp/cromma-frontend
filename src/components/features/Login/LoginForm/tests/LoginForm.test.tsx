import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import LoginForm from '@components/features/Login/LoginForm';
import * as useLoginForm from '@hooks/useLoginForm';

type ReturnType = [boolean, boolean, boolean, (e: React.FormEvent<HTMLFormElement>) => Promise<void>];
describe('LoginForm Component', () => {
	const mockedCloseLoginModal = jest.fn();

	const enterId = (id: string) => {
		const idInput = screen.getByPlaceholderText('id');

		userEvent.type(idInput, id);
	};

	const expectId = (expectedId: string) => {
		const idInput = screen.getByPlaceholderText('id');

		expect(idInput).toHaveValue(expectedId);
	};

	const enterPassword = (password: string) => {
		const passwordInput = screen.getByPlaceholderText('password');

		userEvent.type(passwordInput, password);
	};

	const expectPassword = (expectedPassword: string) => {
		const passwordInput = screen.getByPlaceholderText('password');

		expect(passwordInput).toHaveValue(expectedPassword);
	};

	const clickSubmitButton = () => {
		const submitButton = screen.getByRole('button', {
			name: '로그인',
		});

		userEvent.click(submitButton);
	};

	const checkIsIdInputNotValid = (id: string) => {
		enterId(id);

		clickSubmitButton();

		expect(screen.getByText('아이디는 4글자 이상 10글자 이하의 영어와 숫자 조합이어야 합니다.')).toBeInTheDocument();
	};

	const checkIsPasswordInputNotValid = (password: string) => {
		enterPassword(password);

		clickSubmitButton();

		expect(screen.getByText('비밀번호는 4글자 이상 10글자 이하의 영어와 숫자 조합이어야 합니다.')).toBeInTheDocument();
	};

	const mockedSubmitLoginForm = jest.fn();

	const spyOnUseGetAllPosts = (returnData: ReturnType) => {
		const mockedUseLoginForm = jest.spyOn(useLoginForm, 'default');
		mockedUseLoginForm.mockImplementation(() => returnData);

		render(<LoginForm closeLoginModal={mockedCloseLoginModal} />);
	};

	it('Snapshot', () => {
		const component = renderer.create(<LoginForm closeLoginModal={mockedCloseLoginModal} />);
		expect(component.toJSON()).toMatchSnapshot();
	});

	describe('Init and change inputs', () => {
		beforeEach(() => {
			render(<LoginForm closeLoginModal={mockedCloseLoginModal} />);
		});

		it("Init ID and password with ''", () => {
			expectId('');

			expectPassword('');
		});

		it('Change value when enter input', () => {
			enterId('test@gmail.com');

			expectId('test@gmail.com');

			enterPassword('test1234');

			expectPassword('test1234');
		});
	});

	describe('Show alert message when ID or password is not valid', () => {
		beforeEach(() => {
			render(<LoginForm closeLoginModal={mockedCloseLoginModal} />);
		});

		it('ID must be at least 4 letters and less than 10 letters', () => {
			checkIsIdInputNotValid('tes');

			checkIsIdInputNotValid('testtesttest');
		});

		it('ID must consist of English or numbers', () => {
			checkIsIdInputNotValid('test!!');
		});

		it('Password must be at least 4 letters and less than 10 letters', () => {
			checkIsPasswordInputNotValid('tes');

			checkIsPasswordInputNotValid('testtesttest');
		});

		it('Password must consist of English or numbers', () => {
			checkIsPasswordInputNotValid('test!!');
		});
	});

	it('Call submitLoginForm() when click submit button', () => {
		spyOnUseGetAllPosts([true, true, true, mockedSubmitLoginForm]);

		clickSubmitButton();

		expect(mockedSubmitLoginForm).toHaveBeenCalled();
	});
});
