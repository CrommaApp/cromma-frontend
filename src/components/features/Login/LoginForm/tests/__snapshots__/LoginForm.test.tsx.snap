// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Component Snapshot 1`] = `
<form
  className="sc-bdvvtL jPHANa"
  onSubmit={[Function]}
>
  <ul>
    <li
      className="sc-gsDKAQ ifOJJU"
    >
      <label
        className="a11y-hidden"
        htmlFor="login_input_id"
      >
        ID
      </label>
      <input
        aria-errormessage="id_error_msg"
        aria-invalid={false}
        aria-required="true"
        id="login_input_id"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="id"
        type="text"
        value=""
      />
    </li>
    <li
      className="sc-gsDKAQ ifOJJU"
    >
      <label
        className="a11y-hidden"
        htmlFor="login_input_password"
      >
        Password
      </label>
      <input
        aria-errormessage="password_error_msg"
        aria-invalid={false}
        aria-required="true"
        id="login_input_password"
        onChange={[Function]}
        placeholder="password"
        type="password"
        value=""
      />
    </li>
  </ul>
  <button
    type="submit"
  >
    로그인
  </button>
</form>
`;
